### PR TITLE
Adds configuration option for large file warning threshold.

### PR DESCRIPTION
**Pull request by remexre**
See #10086.

### CODE REVIEW

**Comment by nathansobo** [comment]
@50Wliu Yeah I'm cool with it. Makes sense. If someone can verify this works correctly I'm 👍 to merge.

**Comment by 50Wliu** [comment]
I tested this out and it works correctly.  One thing though @remexre...can you please adjust [this spec](https://github.com/atom/atom/blob/f376316d60b7b0e96ceaff8c631ea6458f31907c/spec/workspace-spec.coffee#L435) to respect the new config option and also to test out a different value to make sure that the option doesn't regress?  Thanks!

**Comment by remexre** [comment]
Like so?

**Comment by nathansobo** [comment]
@remexre I don't think that's quite it. What we need is a test that the limit is configurable. So you would assign a limit to the configuration in the test to a number below the mocked file size and confirm we show the dialog, then reassign the limit to a number above the mocked file size and confirm we don't show the dialog.

**Comment by remexre** [comment]
Is this good?

File: spec/workspace-spec.coffee [hunk 1]
```diff
@@ -432,31 +432,37 @@ describe "Workspace", ->
...
-
+      it "prompts the user to make sure they want to open a file this big", ->
+        atom.config.set "core.warnOnLargeFileLimit", 20
+        test 20, true
+      it "doesn't prompt on files below the limit", ->
+        atom.config.set "core.warnOnLargeFileLimit", 30
+        test 20, false
+      it "prompts for smaller files with a lower limit", ->
+        atom.config.set "core.warnOnLargeFileLimit", 5
+        test 10, true
```
Comment (nathansobo) [inline spec/workspace-spec.coffee]: Can you space this stuff out?

File: spec/workspace-spec.coffee [hunk 2]
```diff
@@ -432,31 +432,37 @@ describe "Workspace", ->
         runs ->
           expect(editor.largeFileMode).toBe true
 
-    describe "when the file is over 20MB", ->
-      it "prompts the user to make sure they want to open a file this big", ->
-        spyOn(fs, 'getSizeSync').andReturn 20 * 1048577 # 20MB
+    describe "when the file is over user-defined limit", ->
+      test = (size, shouldPrompt) ->
```
Comment (nathansobo) [inline spec/workspace-spec.coffee]: Can you give this helper a more descriptive name?

File: spec/workspace-spec.coffee [hunk 3]
```diff
@@ -432,31 +432,37 @@ describe "Workspace", ->
...
+        spyOn(fs, 'getSizeSync').andReturn size * 1048577
         atom.applicationDelegate.confirm.andCallFake -> selectedButtonIndex
         atom.applicationDelegate.confirm()
-        selectedButtonIndex = 1 # cancel
-
         editor = null
         waitsForPromise ->
           workspace.open('sample.js').then (e) -> editor = e
+        if shouldPrompt
```
Comment (nathansobo) [inline spec/workspace-spec.coffee]: If `shouldPrompt` is false, we should assert that it didn't prompt.

**Comment by nathansobo** [comment]
Getting close. I left a few more comments.

File: spec/workspace-spec.coffee (hunk 3 above)
Comment (remexre) [inline spec/workspace-spec.coffee]: How would I do that, exactly? I'm not 100% familiar with Atom's internal structure...

**Comment by remexre** [comment]
First two done, not 100% sure how to do the last one.

**Comment by nathansobo** [comment]
You can also do a negative assertion on your mock...

```coffee
expect(atom.applicationDelegate.confirm).not.toHaveBeenCalled()
```

**Comment by remexre** [comment]
I think it's fixed now? For some reason, `script/test` fails, claiming it doesn't have dpkg, so local tests aren't working for me...

**Comment by 50Wliu** [comment]
@remexre You can view specs by clicking on the `ci/circleci` check.  Right now there's just an issue where `selectedButtonIndex` is undefined.  In addition, I don't think there will be an available application delegate to `confirm()` if the file size is beneath the limit?

**Comment by 50Wliu** [comment]
This looks good now - thanks!

**Comment by nathansobo** [comment]
⚡️ Thanks @remexre!

**Comment by Ben3eeE** [comment]
Closes https://github.com/atom/tree-view/issues/810?

**Comment by remexre** [comment]
I believe so.

**Comment by adambuczynski** [comment]
@remexre does this apply to actually opening files (e.g. double click), or also when previewing files from tree view (single click)? I just had the editor hang on me because I accidentally clicked on a file with 70k+ lines.

**Comment by remexre** [comment]
IIRC previewing

**Comment by adambuczynski** [comment]
Great, thanks

**Pull request merged**